docs/concurrency-adoption-guidelines.md

```diff
@@ -0,0 +1,201 @@
+
+Swift Concurrency is a multi-year effort. It is very valuable for the server community to participate in this multi-year adoption of the concurrency features, one by one, and provide feedback while doing so. As such, we should not hold off adopting concurrency features until Swift 6 as we may miss valuable opportunity to improve the concurrency model.
+
+In 2021 we saw structured concurrency and actors arrive with Swift 5.5. Now is a great time to provide APIs using those primitives. In the future we will see fully checked Swift concurrency. This will come with breaking changes. For this reason adopting the new concurrency features can be split into two phases.
+
+
+## What you can do right now
+
+### `#if` guarding code using Concurrency
+
+In order to have code using concurrency along with code not using concurrency, you may have to `#if` guard certain pieces of code. The correct way to do so is the following:
+
+```swift
+#if compiler(>=5.5) && canImport(_Concurrency)
+...
+#endif
+```
+
+Please note that you do _not_ need to _import_ the `_Concurrency` at all, if it is present it is imported automatically.
+
+```swift
+// DO NOT DO THIS.
+// Instead don't do any import and it'll import automatically when possible.
+#if compiler(>=5.5) && canImport(_Concurrency)
+import _Concurrency
+#endif
+```
+
+
+
+### API Design
+
+Firstly, existing libraries should strive to add `async` functions where possible to their user-facing “surface” APIs in addition to existing `*Future` based APIs wherever possible. These additive APIs can be gated on the Swift version and can be added without breaking existing users' code, for example like this:
+
+```swift
+extension Worker {
+  func work() -> EventLoopFuture<Value> { ... }
+  
+  #if swift(>=5.5)
+  func work() async throws -> Value { ... }
+  #endif
+}
+```
+
+If a function cannot fail but was using futures before, it should not include the `throws` keyword in its new incarnation. 
+
+Such adoption can begin immediately, and should not cause any issues to existing users of existing libraries. 
+
+### SwiftNIO helper functions
+
+To allow an easy transition to async code, SwiftNIO offers a number of helper methods on `EventLoopFuture` and `-Promise`. Those live in the `_NIOConcurrency` module and will move to `NIOCore` once Swift concurrency is released.
+
+On every `EventLoopFuture` you can call `.get()` to transition the future into an `await`-able invocation. If you want to translate async/await calls to an `EventLoopFuture` we recommend the following pattern:
+
+```swift 
+#if swift(>=5.5)
+func yourAsyncFunctionConvertedToAFuture(on eventLoop: EventLoop) 
+    -> EventLoopFuture<Result> {
+    let promise = context.eventLoop.makePromise(of: Out.self)
+    promise.completeWithTask {
+        try await yourMethod(yourInputs)
```

Comment:
Question about this; from what I can tell, this won't run `yourMethod` on an `EventLoop`; (assuming because I tried an example and `MultiThreadedEventLoopGroup.currentEventLoop` is nil).

Is there a way to have async tasks run on an event loop?

I ask since I thought that ideally during a request's lifecycle it won't be hopping off the EventLoop it's received on, even if running async tasks like a database lookup.

I might be confused on what's going on here too 🙂

Reply:
> Is there a way to have async tasks run on an event loop?

Not today; that's what upcoming work on custom executors in Swift Concurrency would allow.

Reply:
The specific use case is that when running database queries, I've got a dict of connection pools, one for each `EventLoop`.

When a query is made, a pool is found using the current event loop's objectidentifier. But it seems like with async functions, they aren't running on an event loop and so if they need to run a database query it's not possible to look up a pool based on that.

Reply:
that's correct